src/Analyzer/Pass/Expression/FunctionCall/FunctionStringFormater.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace PHPSA\Analyzer\Pass\Expression\FunctionCall;
+
+use PHPSA\Context;
+use PhpParser\Node;
+use PhpParser\Node\Stmt;
+use PHPSA\Analyzer\Pass;
+use PhpParser\Node\Expr\Array_;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Scalar\String_;
+use PHPSA\Analyzer\Helper\DefaultMetadataPassTrait;
+
+class FunctionStringFormater extends AbstractFunctionCallAnalyzer
+{
+    use DefaultMetadataPassTrait;
+
+    const DESCRIPTION = 'Format string has same number of placeholders as parameters are passed into and forbid invalid type formats.';
+
+    /**
+     * @var array functions
+     */
+    protected static $functions = [
+        'printf' => 'printf',
+        'sprintf' => 'sprintf'
+    ];
+    /**
+     * Placeholders for type format
+     * @var array
+     */
+    protected $placeholders = [];
+
+    /**
+     * @param FuncCall $funcCall
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(FuncCall $funcCall, Context $context)
+    {
+        $functionName = $this->resolveFunctionName($funcCall, $context);
+        if ($functionName && isset(self::$functions[$functionName])) {
+            if ($funcCall->args) {
+                $args = $funcCall->args;
+
+                if (! ($args[0]->value instanceof String_)) {
+                    $context->notice(
+                        'function_argument_invalid',
+                        sprintf('First parameter of %s must be string', $functionName),
+                        $funcCall
+                    );
+                }
+
+                if (($args[0]->value instanceof String_)) {
+                    $string = $args[0]->value->value;
+                    // get invalid placeholders
+                    preg_match_all("/%[^bcdeEfFgGosuxX]/", $string, $this->placeholders);
```

Comment:
%% is valid and will print a single %, but this would be flagged invalid here.
%%a would be valid too because this would output %a (the regex below for valid placeholders would use %%b as %b right now)

Reply:
Both regex are improved and they are matching properly valid and invalid formats.